rust/src/array.rs

```diff
@@ -18,447 +18,562 @@
 ///! Array types
 use std::any::Any;
 use std::convert::From;
-use std::ops::Add;
+use std::marker::PhantomData;
+use std::mem;
 use std::sync::Arc;
-use std::str;
-use std::string::String;
 
-use super::bitmap::Bitmap;
+use super::array_data::*;
 use super::buffer::*;
-use super::builder::*;
 use super::datatypes::*;
-use super::list::*;
-use super::list_builder::*;
+use util::bit_util;
 
-/// Trait for dealing with different types of Array at runtime when the type of the
+/// Trait for dealing with different types of array at runtime when the type of the
 /// array is not known in advance
 pub trait Array: Send + Sync {
-    /// Returns the length of the array (number of items in the array)
-    fn len(&self) -> usize;
-    /// Returns the number of null values in the array
-    fn null_count(&self) -> usize;
-    /// Optional validity bitmap (can be None if there are no null values)
-    fn validity_bitmap(&self) -> &Option<Bitmap>;
-    /// Return the array as Any so that it can be downcast to a specific implementation
+    /// Returns the array as Any so that it can be downcast to a specific implementation
     fn as_any(&self) -> &Any;
-}
 
-/// Array of List<T>
-pub struct ListArray<T: ArrowPrimitiveType> {
-    len: usize,
-    data: List<T>,
-    null_count: usize,
-    validity_bitmap: Option<Bitmap>,
+    /// Returns a reference-counted pointer to the data of this array
+    fn data(&self) -> ArrayDataRef;
+
+    /// Returns a borrowed & reference-counted pointer to the data of this array
+    fn data_ref(&self) -> &ArrayDataRef;
 }
 
-impl<T> ListArray<T>
-where
-    T: ArrowPrimitiveType,
-{
-    pub fn len(&self) -> usize {
-        self.len
+pub type ArrayRef = Arc<Array>;
+
+/// Constructs an array using the input `data`. Returns a reference-counted `Array`
+/// instance.
+fn make_array(data: ArrayDataRef) -> ArrayRef {
+    // TODO: refactor the DataType enum and remove the clone here.
```

Comment:
Can you explain what you mean by "clone"?

Reply:
Sorry this comment is a bit confusing. I meant the clone on `DataType` which will have to copy all the children when it is a nested type. A better way might be to declare a enum that only represents the "tag" of the type, similar to [`Type`](https://github.com/apache/arrow/blob/master/cpp/src/arrow/type.h#L45) in the C++ version.